### ISSUE TITLE
How to group jobs

### ISSUE THREAD

**Issue by Kunze** [distribworks/dkron#542]
I would like to know if there is a way to group jobs, like a custom tag or something.
My requirement is that I need to create several jobs for an user, and at some point I need to delete all jobs which were not executed.
I tried to do it using "tags" (tags: { userId: "123" }) but the http executor does not run, although the job is created.

**Comment by yvanoers** [comment]
@kunze A job's `owner` property may be useful for your specific use case.

Tags are meant to target specific dkron nodes (https://dkron.io/usage/target-nodes-spec/) to run a job on. If you have no nodes configured with tag `userId: 123` (e.g. https://dkron.io/basics/configuration/), the job will indeed not run.

In a broader sense, there is no grouping or categorization functionality in dkron AFAIK. But in feature request #535 a metadata field has been suggested to provide a solution for this.

**Comment by victorcoder** [comment]
@Kunze There's a feature barely documented that allows you to get jobs from the API filtering by tags https://dkron.io/api/#/jobs/getJobs 

With that list you can filter not executed jobs and delete them.

**Comment by Kunze** [comment]
> I would like to know if there is a way to group jobs, like a custom tag or something.
> My requirement is that I need to create several jobs for an user, and at some point I need to delete all jobs which were not executed.
> I tried to do it using "tags" (tags: { userId: "123" }) but the http executor does not run, although the job is created.

Hi @yvanoers , the problem with "owner" is that I can't filter by that field, currently only filtering by tags is allowed as I can see in the documentation.

**Comment by Kunze** [comment]
> @Kunze There's a feature barely documented that allows you to get jobs from the API filtering by tags https://dkron.io/api/#/jobs/getJobs
> 
> With that list you can filter not executed jobs and delete them.

Hi @victorcoder . Filtering by jobs not executed is not enough because it would bring jobs for all users, and what I need is to filter by a specific user.
Basically what I do is:
- When the user is created, 5 jobs are created (with schedules for 12h, 24h, 48h, 72h, 96h)
- When the user complete its registration, jobs should be deleted (or disabled).

Currently I'm creating jobs using a schedule config in the name of the job, like "send_notification_12h", "send_notification_24h", and then I'm able to delete it by calling DELETE /jobs/{jobName}. 
The problem with that approach is that if someone change the schedule config to 12h, 24h, 48h, 72h (removing 96h), the job named "send_notification_96" will not be deleted when the user complete its registration. So creating jobs with a custom tag (userId: 12345) and getting all jobs with that custom tag would allow me to delete all jobs related to the user.

**Comment by victorcoder** [comment]
Hi @Kunze I think you misread me, you can do that filtering, take a look at https://dkron.io/api/#/jobs/getJobs you can pass a param to that API endpoint indicating the tags for jobs you want to filter by.

**Comment by Kunze** [comment]
Hi @victorcoder 
I'm able to create jobs with tags and filter it using the api, but the problem is that the job does not run when I use custom tags like:
```
"tags": {
     "userId":"12345"
},
```

**Comment by victorcoder** [comment]
@Kunze that's a different issue, if the job has no tags it'll run in all nodes, but as you are adding a tag it will try to run in nodes with that tag. What you can do is to assign the job an extra tag that matches a node, for example: `dkron_server: true` or any other.

Then the user tag will be _ignored_ and it will use the actual matching tag.

**Comment by Kunze** [comment]
@victorcoder I tried it but the job with tags does not run:

- this jobs runs when I click on "play" 
```
{
  "name": "test-with-no-tags",
  "schedule": "@at 2019-06-06T19:10:27-03:00",
  "executor": "shell",
  "executor_config": {
     "command":"echo test"
  }
}
```

-this job does not run when I click on "play"
```
{
  "name": "test-with-tags",
  "tags": {
    "dkron_server": "true",
    "userId":"12345"	
  },
  "schedule": "@at 2019-06-06T19:10:27-03:00",
  "executor": "shell",
  "executor_config": {
    "command":"echo test"
  }
}
```
![dkron](https://user-images.githubusercontent.com/1281157/59211993-a9f8a000-8b87-11e9-9231-9c9d059083ec.PNG)

Am I doing something wrong?

**Comment by yvanoers** [comment]
@Kunze Could you also share your dkron server configuration?

**Comment by Kunze** [comment]
> @Kunze Could you also share your dkron server configuration?

Hi @yvanoers , I didn't change anything in configuration, so I'm using the default.

**Comment by victorcoder** [comment]
Sorry @Kunze my mistake, it won't run if not all tags matches. I'm thinking in a workaround for this, I'll get back with something

**Comment by victorcoder** [comment]
It is not possible to make this work without modifications. I've been thinking in a simple hack, would something like this work for you @Kunze?

`"userId": "12345:0"`

I can treat this as an _ignorable tag_  discarding it when filtering nodes. It would just complicate querying the API a bit more.

**Comment by yvanoers** [comment]
@victorcoder, may I suggest implementation of proper filtering on the `/jobs` collection instead?
Perhaps starting with single value fields such as `owner`, allowing queries like `/jobs?filter=owner eq 'someid'`.
It may be more work than the hack, but it would lay the groundwork for more filters and isn't, well, a hack.

**Comment by victorcoder** [comment]
@yvanoers as talked in the related issue https://github.com/victorcoder/dkron/issues/535 with @viperfx and @fopina, the idea of adding a new `metadata` set of tags would add flexibility for filtering jobs in the API/UI and do not interfere with the `tags`.

Note: this feature will be only added to v2 branch, I'm not adding more features to v1, only bug fixes.

**Comment by yvanoers** [comment]
@victorcoder Good to know about the v2 branch, thanks.

The metadata feature from #535 is, as I understand it, only about adding the field. And that would allow for filtering by the API <i>client</i>. What I gather from @kunze is the requirement to have the filtering done by the <i>server</i>. Perhaps a more generic filtering mechanism is already in v2, I'm not aware (yet). If so, please disregard my suggestion.

**Comment by victorcoder** [comment]
@yvanoers you can currently filter job by tag from the _server_. The new metadata field would allow to also filter jobs by the _server_.

**Comment by victorcoder** [comment]
Fixed in #543
